### PR TITLE
ipa_pwd.c: Use OpenSSL instead of NSS for hashing 

### PR DESCRIPTION
**Pull request by Carbenium**
Fixes: https://pagure.io/freeipa/issue/6857

### CODE REVIEW

File: util/ipa_pwd.c [hunk 1]
```diff
@@ -162,30 +161,30 @@ static int ipapwd_gentime_cmp(const void *p1, const void *p2)
 #define SHA_SALT_LENGTH 8
 
 /* SHA*_LENGTH leghts come from nss3/hasht.h */
-#define SHA_HASH_MAX_LENGTH SHA512_LENGTH
+#define SHA_HASH_MAX_LENGTH SHA512_DIGEST_LENGTH
```
[the remark this answers is not in the document]
Comment (Carbenium) [reply util/ipa_pwd.c]: Done.

File: util/ipa_pwd.c [hunk 2]
```diff
@@ -307,35 +304,42 @@ static int ipapwd_cmp_password(char *password, char *historyString)
     b64part++;
     b64_len = strlen(b64part);
 
-    item = NSSBase64_DecodeBuffer(NULL, NULL, b64part, b64_len);
-    if (!item) {
+    char *item_data = (char *) malloc(sizeof(char) * b64_len);
```
[the remark this answers is not in the document]
Comment (Carbenium) [reply util/ipa_pwd.c]: Done.

File: util/ipa_pwd.c [hunk 3]
```diff
@@ -307,35 +304,42 @@ static int ipapwd_cmp_password(char *password, char *historyString)
...
         ret = 1;
         goto done;
     }
 
     ret = 0;
 
 done:
-    SECITEM_FreeItem(item, 1);
+    if (item_data)
+        free(item_data);
```
[the remark this answers is not in the document]
Comment (Carbenium) [reply util/ipa_pwd.c]: Done.

**Comment by Carbenium** [comment]
@tiran I left the changes in a separate commit for easier review. If you're fine with them, I'm going to squash it.

File: util/ipa_pwd.c [hunk 4]
```diff
@@ -162,30 +161,30 @@ static int ipapwd_gentime_cmp(const void *p1, const void *p2)
...
 static int ipapwd_hash_type_to_alg(char *hash_type,
-                                   SECOidTag *hash_alg,
+                                   const EVP_MD **hash_alg,
                                    unsigned int *hash_alg_len)
 {
     if (strncmp("{SSHA}", hash_type, 6) == 0) {
-        *hash_alg = SEC_OID_SHA1;
-        *hash_alg_len = SHA1_LENGTH;
+        *hash_alg = EVP_sha1();
+        *hash_alg_len = SHA_DIGEST_LENGTH;
```
[the remark this answers is not in the document]
Comment (Carbenium) [reply util/ipa_pwd.c]: Done.

File: util/ipa_pwd.c [hunk 5]
```diff
@@ -307,35 +304,42 @@ static int ipapwd_cmp_password(char *password, char *historyString)
...
     }
 
-    if (hash_len != item->len) {
+    if (hash_len != item_len) {
         ret = 1;
         goto done;
     }
 
-    if (memcmp(item->data, hash, hash_len)) {
+    if (memcmp(item_data, hash, hash_len)) {
```
[the remark this answers is not in the document]
Comment (Carbenium) [reply util/ipa_pwd.c]: Done.

**Comment by abbra** [comment]
@Carbenium please squash the fixes. They all look good.

**Comment by Carbenium** [comment]
@abbra Done.

**Comment by abbra** [comment]
master:

* f2d854886fdaa4d61965c1af6c02060d816a1fbf util: add unit test for pw hashing
* 68af4f39c19fb52fcf6ed674b883ee30026bdb4b util: replace NSS usage with OpenSSL

**Pull request closed without merging**